### DISCUSSION TITLE
How does trpc know which invocations are on the same batch?

### DISCUSSION THREAD

**Discussion by ylc395** [trpc/trpc#5306] [General]
According to doc:

> After that, you can make use of batching by setting all your procedures in a Promise.all. The code below will produce exactly one HTTP request and on the server exactly one database query:

This feature is cool, I wonder how trpc knows some invocation are on the same batch? Does trpc rewrite the `Promise.all`?

**Comment by KATT** [comment]
We have a dataloder that waits one tick before doing any calls to the server (when you use the httpBatchLink) and groups everything into 1 req

**Reply by ylc395** [reply]
so all invocations in one tick will be batched?

**Comment by mnpenner** [comment]
The part I want to know about is this:

>  and on the server exactly one database query:

Can we see the implementation of `byId`? I'm assuming `byId` takes a single ID and it would have been typed as such in the `.input` method on the server, so how are these being combined into a single database query?

**Reply by theoludwig** [reply]
Yes. I don't think this statement is true.

It indeed produce exactly one HTTP request, but still nonetheless produce multiple database query.

**Reply by KATT** [reply]
With Prisma you get automatic batching for `findUnique`-queries that solves the so called N+1 problem - using dataloaders work for this too

**Reply by s-h-a-d-o-w** [reply]
Shouldn't the docs then say something like "(If your ORM supports automatic batching for this operation, since the handler will still be called for each payload.)"?

Because reading the tRPC docs, I'm really interested in learning how tRPC works. Which in this case means - will the route handler be called only once with an array or multiple times with each element?
I now get that it's the latter but... it would be great if the docs made that clear.
